### ISSUE TITLE
Make MySQL metadata use `utf8mb4`

### ISSUE THREAD

**Issue by flamber** [metabase/metabase#10691]
With 0.33, it seems like the table joins now requires metadata to be `utf8mb4`, but many older MySQL/MariaDB installations has been setup with various older charsets/collations like `latin1` or `utf8`, which causes errors when saving questions like:
```
ERROR middleware.results-metadata :: Error recording results metadata for query:
java.sql.BatchUpdateException (conn=180) Incorrect string value: ‘\xE2\x86\x92 Fi…’ for column ‘result_metadata’ at row 1
```

There has been a surge in reports in the forum the past couple of weeks.

Ways to check the charset/collation (cannot be executed from Metabase editor):
```SQL
-- Set the database name to the Metabase metadata database - change it to match your db:
USE metabase;
-- Show system variables
SHOW VARIABLES WHERE Variable_name LIKE 'character_set_%' OR Variable_name LIKE 'collation%';
-- Show metadata database/tables/columns
SELECT TABLE_SCHEMA,
       TABLE_NAME,
       CCSA.CHARACTER_SET_NAME AS DEFAULT_CHAR_SET,
       COLUMN_NAME,
       COLUMN_TYPE,
       C.CHARACTER_SET_NAME
  FROM information_schema.TABLES AS T
  JOIN information_schema.COLUMNS AS C USING (TABLE_SCHEMA, TABLE_NAME)
  JOIN information_schema.COLLATION_CHARACTER_SET_APPLICABILITY AS CCSA
       ON (T.TABLE_COLLATION = CCSA.COLLATION_NAME)
 WHERE TABLE_SCHEMA=SCHEMA()
   AND C.DATA_TYPE IN ('enum', 'varchar', 'char', 'text', 'mediumtext', 'longtext')
 ORDER BY TABLE_SCHEMA,
          TABLE_NAME,
          COLUMN_NAME;
```

Could Metabase use `utf8mb4`, when people are migrating? Also setting engine to `innodb`, since some older MySQL installations has default engine as `myisam`.

And/or log a warning on startup for existing installations that are not following the recommendation?

A tiny bit similar to #10365

<details>
<summary>Click here to see the commands for doing the change</summary>

***BE CAREFUL!***
This expects you are using `latin1` currently - change it to match your current charset
And input your credentials here instead of: `XXusernameXX`, `XXpasswordXX` and `XXdatabaseXX`

`mysqldump -uXXusernameXX -pXXpasswordXX -c -e --default-character-set=utf8mb4 --single-transaction --skip-set-charset --add-drop-database -B XXmetabaseXX > dump.sql`

`sed -i.bak -e 's/DEFAULT CHARACTER SET latin1/DEFAULT CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_ci/' -e 's/DEFAULT CHARSET=latin1/DEFAULT CHARSET=utf8mb4/' dump.sql`

`mysql -uXXusernameXX -pXXpasswordXX < dump.sql`
</details>

**Comment by salsakran** [comment]
@flamber is this preventing people from saving questions if they don't have MySQL set to use `utf8mb4`?

We have no such requirement, so erroring out on a save makes this an outright bug. How common is this setting in the wild? Could see this being a P1 if it's common enough.

**Comment by flamber** [comment]
@salsakran 
Yes, it seems so from #10292 and [several forum posts](https://discourse.metabase.com/search?q=utf8mb%20order%3Alatest). And it's also making migration difficult, see #9741 and #9786.

Sadly, I see this bad configuration so many places - along with no timezone (#10365) - and those two combined can mess up the data pretty good. For many years, MySQL was shipping with a really horrible default configuration, which is why this exists so many places.

**Comment by RMMSecurity** [comment]
@salsakran

I think I was referenced by flamber.

This happened on two different installs I had, though resolved now left it completely unusable on those versions. I'd definitely rate it as a P1. The install guides never gave any special instructions on it so I'd venture they used the default charsets/collation.

**Comment by boltronics** [comment]
I've seen this from users - and had to dig through the logs to see why things weren't saving.

I tried converting the DB to utf8mb4, and kept running into:
`ERROR 1071 (42000) at line 30: Specified key was too long; max key length is 767 bytes`

I read on Stack Overflow that adding `SET @@global.innodb_large_prefix = 1;` is a possible work-around, however since I have a number of important databases on the same host, I'd rather avoid that so didn't try it.

**Comment by flamber** [comment]
@boltronics 
Which version of MySQL specifically are you using for metadata?
`innodb-large-prefix` has been deprecated since [5.7](https://dev.mysql.com/doc/refman/5.7/en/innodb-parameters.html#sysvar_innodb_large_prefix) and [10.2](https://mariadb.com/kb/en/library/innodb-system-variables/#innodb_large_prefix).
The default configuration since 5.7.7 and 10.2.2 is `ON`. Before that the default was `OFF`.
Sounds like one of the main pain points in this old issue #2688, which should probably be closed.

Since it's not database/table/column specific, then I don't see Metabase ever having a solution for that, but a warning in the log might be helpful, as well as the documentation.

**Comment by boltronics** [comment]
@flamber Thanks. We are using MySQL 5.6.34 as provided by Amazon RDS. `innodb_large_prefix` is available but has not been set so would be using the default.
`Dec  2 12:52:58 ip-172-31-20-173 java[8379]: Caused by: java.sql.SQLException: Incorrect string value: '\xE2\x86\x92 Un...' for column 'result_metadata' at row 1`

Seems to be triggered by the special character in `"display_name":"Membership Periods → Until Time"` (which does not look possible to change via the UI). 

The documentation doesn't say anything about needing MySQL 5.7.7+ or needing to tweak the `innodb_large_prefix` setting when upgrading.

**Comment by rthouvenin** [comment]
Migrating the metadata DB and all its tables to utf8mb4 solved the problem for me. However this is a rather tedious process and I think not all Metabase admins might have control over this, so ideally Metabase shouldn't have this requirement.

**Comment by flamber** [comment]
@rthouvenin I would have to disagree. If you don't have control over the database, where Metabase stores its data, then you don't have control over Metabase. I'm not talking about the database server.
Metabase should just do this automatically, when the database is created during first setup/migration.
MySQL's version of `utf8` is broken, which is why `utf8mb4` is the recommended way for UTF-8.

**Comment by rthouvenin** [comment]
@flamber Right, it indeed makes sense that the person deploying Metabase also has control on the Metabase DB. I guess I meant that doing the migration is a rather advanced DB admin skill, which I thought is not expected of Metabase admins. 
Regardless, if Metabase takes care of it transparently then it is not really a requirement, as in it's not something listed as "do this before you install / updgrade", that is what I meant by "it shouldn't have this requirement". 

Looking forward to seeing this solved as the migration I did was on one test instance but I take care of other instances that will face this problem when I upgrade to 0.33.

**Comment by camsaul** [comment]
Does anyone have any suggestions for how we should go about doing this in Metabase itself? Are there SQL statements we could run for people using MySQL as part of a DB migration that would fix this issue? Running external `mysql` commands isn't something that is really possible from Metabase itself. Not all systems have the MySQL CLI installed.

**Comment by camsaul** [comment]
@flamber (or anyone else who has more experience with legacy MySQL), since we're not really in charge of creating the Database, would you recommend I alter every 
```sql
CREATE TABLE ...
``` 
statement for MySQL to 

```sql
CREATE TABLE ... 
ENGINE InnoDB 
CHARACTER SET utf8mb4 
COLLATE utf8mb4_unicode_ci
```

would that do the trick?

**Comment by camsaul** [comment]
Fixed by #11742. Please note that the fix will only affect newly-created tables; for existing tables you can manually convert them using the commands posted elsewhere in this issue. 

If you were unable to get Metabase up and running (i.e., you don't have important data in your MySQL application database) it would probably be fastest to either recreate the database or drop all the tables in it so they will get recreated correctly.

**Comment by salsakran** [comment]
@camsaul does this mean that people currently running into this will need to manually run commands against their application database to get things working?

**Comment by camsaul** [comment]
@salsakran no, it only means anyone who ran into this bug would have to. As far as I can tell, this bug only affected a small subset of our MySQL users. 

The PR explicitly specifies the character encoding and engine MySQL should use for Metabase application DB tables instead of relying on whatever defaults are configured. 

InnoDB is already the default engine in MySQL 5.7+, and we've never officially supported anything below MySQL 5.7. We have had anecdotal reports of of people using Metabase with MySQL 5.6, but we've never tested against it. So this change would only affect people using versions of MySQL we don't officially support.

I'm trying to work out at what point `utf8mb4` became the default character set, but I've tried MySQL 5.7.x on maybe 6 different computers (installing via package manager) and it was already the default on all of those. So the character set change will only affect older MySQL instances.

**Comment by camsaul** [comment]
I think I've worked out a way to do conversion automatically in a DB migration

**Comment by camsaul** [comment]
@flamber @pubmania @RMMSecurity @alejandrorial @moret @eschultze @rthouvenin @alantygel @rthouvenin @boltronics @huseinmb @alpha1 @Microsheep @Shra25 @etiennecallies (at-mentioning you because you either opened a related issue, commented on one, or thumbs-upped one):

A quick status update on this issue. In PR #11742 (already merged) I tweaked the way we run database migrations to default to InnoDB/`utf8mb4` for all newly created Tables. That means any new instances of Metabase or instances where you migrate from H2 to MySQL or MariaDB won't run into this problem. I've been considering how to address this issue for existing users. 

In PR #11753 (still open) I've added logic to automatically migrate the application DB to `utf8mb4` if needed when Metabase launches. I could really use some help from a few people testing it to make sure it works as expected. Any chance you can give it a try with a Metabase application DB that's not already `utf8mb4`?

**Comment by rthouvenin** [comment]
@camsaul I will hopefully give it a try tomorrow

**Comment by rthouvenin** [comment]
@camsaul When I check out your branch and build it with `bin/build no-translations`, I get this error after some time:

> bin/build: line 24: 13989 Killed                  NODE_ENV=production ./node_modules/.bin/webpack --bail

The same command completes successfully from commit tagged v0.33.6

Let me know if there is a better place to discuss this issue.

**Comment by camsaul** [comment]
@rthouvenin  I've build a JAR that you can use to test it:

https://downloads.metabase.com/mysql-utf8-fix/metabase.jar

SHA256 sum:

```
97d911e80867cca5ab8b03d9a1eec9ea9e8f741816999bb094bc960eb47e513b
```

**Comment by rthouvenin** [comment]
Looks good to me!

I ran 2 tests:

 - Initialized an empty DB with v0.33.6, checked that the charset was not utf8mb4 and that I could not save a question that groups by a joined field. Then ran the jar you provided on same DB, checked that migrations happened and charset of tables was utf8mb4 and that I could save the same question.
 - Then ran same test on the copy of a real DB coming from an instance running 0.32.x with a SQL dump of 217MB.

Both tests were successful. The migration of the real DB took less than 30s on a VPS with 2 CPU and 4GB RAM, half of that time was spent on the table `query_execution`.

**Comment by camsaul** [comment]
Thanks @rthouvenin, that's really helpful!

**Comment by camsaul** [comment]
Fixed by #11753
